docs/distributed/distributed_autofaiss.md

```diff
@@ -0,0 +1,222 @@
+# distributed autofaiss
+
+If you want to generate an index from billion of embeddings, this guide is for you.
+
+This guide is about using pyspark to run autofaiss in multiple nodes.
+
+you may also be interested by [distributed img2dataset](https://github.com/rom1504/img2dataset/blob/main/examples/distributed_img2dataset_tutorial.md)
+and [distributed clip inference](https://github.com/rom1504/clip-retrieval/blob/main/docs/distributed_clip_inference.md)
+
+We will be assuming ubuntu 20.04.
+
+## Setup the master node
+
+On the master node:
+
+First download spark:
+```bash
+wget https://archive.apache.org/dist/spark/spark-3.2.1/spark-3.2.1-bin-hadoop3.2.tgz
+tar xf spark-3.2.1-bin-hadoop3.2.tgz
+```
+
+Then download autofaiss:
+```bash
+rm -rf autofaiss.pex
+wget https://github.com/rom1504/autofaiss/releases/latest/download/autofaiss-3.8.pex -O autofaiss.pex
+chmod +x autofaiss.pex
+```
+
+If the master node cannot open ports that are visible from your local machine, you can do a tunnel between your local machine and the master node to be able to see the spark ui (at http://localhost:8080)
+```bash
+ssh -L 8080:localhost:8080 -L 4040:localhost:4040 master_node
+```
+
+
+## Setup the worker nodes
+
+### ssh basic setup
+
+Still in the master node, create a ips.txt with the ips of all the nodes
+
+```bash
+ssh-keyscan `cat ips.txt` >> ~/.ssh/known_hosts
+```
+
+You may use a script like this to fill your .ssh/config file
+```
+def generate(ip):
+    print(
+        f"Host {ip}\n"
+        f"        HostName {ip}\n"
+        "        User ubuntu\n"
+        "        IdentityFile ~/yourkey.pem"
+        )
+
+with open("ips.txt") as f:
+    lines = f.readlines()
+    for line in lines:
+        generate(line.strip())
+```
+python3 generate.py >> ~/.ssh/config
+
+Install pssh with `sudo apt install pssh`
+
+Pick the right username (USER) for the worker nodes, then run this to check your parallel ssh setup:
+```bash
+USER=ubuntu
+```
+
+Optionally, if another node than the current one has access to the worker nodes, you may need to add a ssh key to all the nodes with:
+```
+for IP in `cat ips.txt`
+do
+        ssh-copy-id -i the_new_id_rsa $USER@$IP
+done
+```
+
+Check you can connect to all the nodes with:
+```
+parallel-ssh -l $USER -i -h  ips.txt uname -a
+```
+
+##### Install some packages
+
+```bash
+parallel-ssh -l $USER -i -h  ips.txt "sudo apt update"
+parallel-ssh -l $USER -i -h  ips.txt "sudo apt install openjdk-11-jre-headless libgl1 htop tmux bwm-ng sshfs python3-distutils python3-apt python3.8 -y"
+```
+
+
+#### [Optional] Network setting on aws
+
+put in same VPC and security group and allow inbound
+
+##### Download img2dataset on all nodes
```

Comment:
The size of the title is too small. And, I guess you mean autofaiss instead of img2dataset.

Reply:
fixed